File: ruby/A58.rb
# 58
# セグメント木(RMQ)
# 実行時間: 3s以内

start_time = Time.now

class SegmentTree
  attr_accessor :leaf_size, :operator, :id_elm, :tree

  def initialize(arr, id_elm, &method)
    n = arr.size
    @operator = method
    @id_elm = id_elm
    @leaf_size = 1 << (n - 1).bit_length
    @tree = [id_elm] * leaf_size
    @tree.concat(arr)
    (leaf_size - 1).downto(1) do |i|
      tree[i] = operator.call(tree[i * 2], tree[i * 2 + 1])
    end
  end

  def update(pos, x)
    u = pos + leaf_size - 1
    tree[u] = x
    while u > 1
      u /= 2
      tree[u] = operator.call(tree[u * 2], tree[u * 2 + 1])
    end
  end

  def query(l, r, a, b, u)
    return id_elm if r <= a || b <= l
    return tree[u] if l <= a && b <= r
    mid = (a + b) / 2
    left = query(l, r, a, mid, u * 2)
    right = query(l, r, mid, b, u * 2 + 1)
    operator.call(left, right)
  end
end

File.open("question/A58.txt", "r") do |f|
  N, Q = f.gets.split.map(&:to_i)
  QUERY = Array.new(Q) { f.gets.split.map(&:to_i) }
end

seg_tree = SegmentTree.new([0] * N, -1) {|i, j| [i, j].max }

QUERY.each do |query|
  case query[0]
  when 1
    seg_tree.update(*query[1, 2])
  when 2
    puts seg_tree.query(*query[1, 2], 1, seg_tree.leaf_size + 1, 1)
  end
end

puts "\n処理時間: #{((Time.now - start_time) * 1_000).round(2)} ms"

File: javascript/A58.js
// A58
// セグメント木

class SegmentTree {
  constructor(arg, id_elm, callback) {

    if (Array.isArray(arg)) {
      this.n = arg.length;
    } else if (typeof arg === "number") {
      this.n = arg;
    }
    this.leaf_size = 1;
    while (this.leaf_size < this.n) this.leaf_size <<= 1;

    this.id_elm = id_elm;
    this.ope = callback;
    this.tree = [...Array(2 * this.leaf_size)].map(_ => this.id_elm);

    if (Array.isArray(arg)) {
      arg.forEach((a, i) => this.tree[this.leaf_size + i] = a);
      for (let i = this.leaf_size - 1; i >0; i--) this.#update(i);
    }
  }

  set(pos, value) {
    let idx = this.leaf_size + pos;
    this.tree[idx] = value;
    while (idx > 1) {
      this.#update(idx >>= 1);
    }
  }

  prod(l, r) {
    let left = this.leaf_size + l;
    let right = this.leaf_size + r;
    let res = this.id_elm;
    while (left < right) {
      if (left % 2 === 1) {
        res = this.ope(res, this.tree[left++]);
      }
      if (right % 2 === 1) {
        res = this.ope(res, this.tree[--right]);
      }
      left >>= 1;
      right >>= 1;
    }
    return res;
  }

  #update(idx) {
    this.tree[idx] = this.ope(this.tree[2 * idx], this.tree[2 * idx + 1]);
  }
}

const input = require("fs").readFileSync("/dev/stdin", "utf8").split("\n");
let n = 0;
const [N, Q] = input[n++].split(" ").map(Number);
const st = new SegmentTree(N, 0, Math.max);
const ans = [];

for (let i = 0; i < Q; i++) {
  const [j, l, r] = input[n++].split(" ").map(Number);
  switch(j) {
    case 1:
      st.set(l - 1, r)
    break;
    case 2:
      ans.push(st.prod(l - 1, r - 1));
    break;
  }
}

console.log(ans.join("\n"));
